README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/jadiunr/check-load)
![Go Test](https://github.com/jadiunr/check-load/workflows/Go%20Test/badge.svg)
![goreleaser](https://github.com/jadiunr/check-load/workflows/goreleaser/badge.svg)

# Sensu load average check

## Table of Contents
- [Overview](#overview)
- [Usage examples](#usage-examples)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Check definition](#check-definition)
- [Installation from source](#installation-from-source)
- [Contributing](#contributing)

## Overview

The Sensu load average check is a [Sensu Check][6] that provides alerting and metrics for load averages. Metrics are provided in [nagios_perfdata](https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/collect-metrics-with-checks/#supported-output-metric-formats) format.

## Usage examples

```
Check load averages and provide metrics

Usage:
  check-load [flags]
  check-load [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -c, --critical string   Critical threshold for load averages (default "0.85,0.8,0.75")
  -h, --help              help for check-load
  -m, --metricsonly       Outputs only the metrics without checking the threshold.
  -r, --percpu            Check the load average divided by the number of CPUs.
  -w, --warning string    Warning threshold for load averages (default "0.75,0.7,0.65")

Use "check-load [command] --help" for more information about a command.
```

## Configuration

### Asset registration

[Sensu Assets][10] are the best way to make use of this plugin. If you're not using an asset, please
consider doing so! If you're using sensuctl 5.13 with Sensu Backend 5.13 or later, you can use the
following command to add the asset:

```
sensuctl asset add jadiunr/check-load
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index](https://bonsai.sensu.io/assets/jadiunr/check-load).

### Check definition

```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: check-load
  namespace: default
spec:
  command: check-load -r -w 0.8,0.7,0.6 -c 0.9,0.8,0.7
  output_metric_format: nagios_perfdata
  output_metric_handlers:
  - influxdb
  subscriptions:
  - system
  runtime_assets:
  - jadiunr/check-load
  interval: 60
  publish: true
```

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an Asset. If you would
like to compile and install the plugin from source or contribute to it, download the latest version
or create an executable script from this source.

From the local path of the check-load repository:

```
go build
```

## Contributing

For more information about contributing to this plugin, see [Contributing][1].

[1]: https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md
[2]: https://github.com/sensu-community/sensu-plugin-sdk
[3]: https://github.com/sensu-plugins/community/blob/master/PLUGIN_STYLEGUIDE.md
[4]: https://github.com/sensu-community/check-plugin-template/blob/master/.github/workflows/release.yml
[5]: https://github.com/sensu-community/check-plugin-template/actions
[6]: https://docs.sensu.io/sensu-go/latest/reference/checks/
[7]: https://github.com/sensu-community/check-plugin-template/blob/master/main.go
[8]: https://bonsai.sensu.io/
[9]: https://github.com/sensu-community/sensu-plugin-tool
[10]: https://docs.sensu.io/sensu-go/latest/reference/assets/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.1.0] - 2022-02-18

### Added
- Initial release


main.go
package main

import (
	"fmt"
    "strconv"
    "strings"

	"github.com/sensu/sensu-go/types"
	"github.com/sensu/sensu-plugin-sdk/sensu"
	"github.com/shirou/gopsutil/v3/cpu"
	"github.com/shirou/gopsutil/v3/load"
)

// Config represents the check plugin config.
type Config struct {
	sensu.PluginConfig
	Critical	string
	Warning		string
    PerCPU      bool
	MetricsOnly bool
}

var (
	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "check-load",
			Short:    "Check load averages and provide metrics",
			Keyspace: "sensu.io/plugins/check-load/config",
		},
	}

	options = []*sensu.PluginConfigOption{
		{
			Path:      "critical",
			Argument:  "critical",
			Shorthand: "c",
			Default:   "0.85,0.8,0.75",
			Usage:     "Critical threshold for load averages",
			Value:     &plugin.Critical,
		},
		{
			Path:      "warning",
			Argument:  "warning",
			Shorthand: "w",
			Default:   "0.75,0.7,0.65",
			Usage:     "Warning threshold for load averages",
			Value:     &plugin.Warning,
		},
        {
            Path:      "percpu",
            Argument:  "percpu",
            Shorthand: "r",
            Usage:     "Check the load average divided by the number of CPUs.",
            Value:     &plugin.PerCPU,
        },
		{
			Path:	   "metricsonly",
			Argument:  "metricsonly",
			Shorthand: "m",
			Default:   false,
			Usage:     "Outputs only the metrics without checking the threshold.",
            Value:     &plugin.MetricsOnly,
		},
	}

    cload [3]float64
    wload [3]float64
)

func main() {
	check := sensu.NewGoCheck(&plugin.PluginConfig, options, checkArgs, executeCheck, false)
	check.Execute()
}

func checkArgs(event *types.Event) (int, error) {
    for i, thStr := range strings.Split(plugin.Critical, ",") {
        th, err := strconv.ParseFloat(thStr, 64)
        if err != nil {
            return sensu.CheckStateWarning, fmt.Errorf("Could not parse the critical threshold.")
        }
        cload[i] = th
    }
    for i, thStr := range strings.Split(plugin.Warning, ",") {
        th, err := strconv.ParseFloat(thStr, 64)
        if err != nil {
            return sensu.CheckStateWarning, fmt.Errorf("Could not parse the warning threshold.")
        }
        wload[i] = th
    }
	for i, w := range wload {
        if w > cload[i] {
            return sensu.CheckStateWarning, fmt.Errorf("--warning cannot be greater than --critical")
        }
    }
    return sensu.CheckStateOK, nil
}

func executeCheck(event *types.Event) (int, error) {
	info, err := cpu.Info()
    if err != nil {
        return sensu.CheckStateCritical, fmt.Errorf("Error: obtaining CPU info: %v", err)
    } else if len(info) < 1 {
        return sensu.CheckStateCritical, fmt.Errorf("Error: no CPU info found")
    }
    cores := float64(0)
    for _, i := range info {
        cores += float64(i.Cores)
    }
    var loadStats *load.AvgStat
    for i := 1; i < 10; i++ {
        loadStats, err = load.Avg()
    }
    if err != nil {
        return sensu.CheckStateCritical, fmt.Errorf("Error: obtaining load stats: %v", err)
    }

    load := []float64{loadStats.Load1, loadStats.Load5, loadStats.Load15}
    loadPerCPU := []float64{load[0] / cores, load[1] / cores, load[2] / cores}
    perfData := fmt.Sprintf(
        "load1=%.2f, load5=%.2f, load15=%.2f, load1_per_cpu=%.2f, load5_per_cpu=%.2f, load15_per_cpu=%.2f",
        load[0], load[1], load[2],
        loadPerCPU[0], loadPerCPU[1], loadPerCPU[2],
    )
    var targetLoad = load
    if plugin.PerCPU {
        targetLoad = loadPerCPU
    }
    if !plugin.MetricsOnly {
        for i, l := range targetLoad {
            if l > cload[i] {
                fmt.Printf("%s Critical - load average: %s | %s\n", plugin.PluginConfig.Name, fmt.Sprintf("%.2f, %.2f, %.2f", targetLoad[0], targetLoad[1], targetLoad[2]), perfData)
                return sensu.CheckStateCritical, nil
            }
            if l > wload[i] {
                fmt.Printf("%s Warning - load average: %s | %s\n", plugin.PluginConfig.Name, fmt.Sprintf("%.2f, %.2f, %.2f", targetLoad[0], targetLoad[1], targetLoad[2]), perfData)
                return sensu.CheckStateWarning, nil
            }
        }
    }

    fmt.Printf("%s OK - load average: %s | %s\n", plugin.PluginConfig.Name, fmt.Sprintf("%.2f, %.2f, %.2f", targetLoad[0], targetLoad[1], targetLoad[2]), perfData)
    return sensu.CheckStateOK, nil
}
